libcextract/SymbolExternalizer.cpp

```diff
@@ -772,236 +753,64 @@ void SymbolExternalizer::Rewrite_Macros(std::string const &to_look_for, std::str
-        IdentifierInfo *new_id = AST->getPreprocessor().getIdentifierInfo(new_name);
-        DeclarationName new_decl_name(new_id);
-
-        decl->setDeclName(new_decl_name);
-        new_decl = decl;
-
-        /* Replace text content of old declaration.  */
-        Replace_Text(id_range, new_name, 100);
-
-        must_update = true;
-        wrap = true;
+  /* If the symbol is available in the debuginfo and is an EXTERN symbol, we
+     do not need to rewrite it, but rather we need to erase any declaration
+     with body of it.  */
+  if (IA.Can_Decide_Visibility()) {
+    if (IA.Is_Externally_Visible(to_externalize))
+      return ExternalizationType::WEAK;
 
-        /* Update any macros that may reference the symbol.  */
-        Rewrite_Macros(to_externalize, new_name);
-        externalized = true;
-      }
-    }
+    return ExternalizationType::STRONG;
   }
 
-  return externalized;
+  /* Well, we don't have information so we simply strongly externalize
+     everything.  */
+  return ExternalizationType::STRONG;
 }
 
-void SymbolExternalizer::Externalize_Symbol(const std::string &to_externalize)
+SymbolUpdateStatus *SymbolExternalizer::getSymbolsUpdateStatus(const StringRef &sym)
 {
-  /* If the symbol is available in the debuginfo and is an EXTERN symbol, we
-     do not need to rewrite it, but rather we need to erase any declaration
-     with body of it.  */
-  if (IA.Can_Decide_Visibility()) {
-    if (IA.Is_Externally_Visible(to_externalize)) {
-      Weakly_Externalize_Symbol(to_externalize);
-    } else {
-      Strongly_Externalize_Symbol(to_externalize);
-    }
-  } else {
-    /* Well, we don't have information so we simply strongly externalize
-       everything.  */
-    Strongly_Externalize_Symbol(to_externalize);
-  }
+  auto ret = SymbolsMap.find(sym);
+  if (ret == SymbolsMap.end())
+    return nullptr;
+
+  return &ret->second;
 }
 
-void SymbolExternalizer::Externalize_Symbols(std::vector<std::string> const &to_externalize_array)
+void SymbolExternalizer::Externalize_Symbols(std::vector<std::string> const &to_externalize_array,
+                                              std::vector<std::string> &to_rename_array)
 {
   for (const std::string &to_externalize : to_externalize_array) {
-    Externalize_Symbol(to_externalize);
+    SymbolsMap.insert({to_externalize, {"", Get_Symbol_Ext_Type(to_externalize), false, false}});
   }
-}
 
-void SymbolExternalizer::Rename_Symbols(std::vector<std::string> &to_rename_array)
-{
   for (std::string &to_externalize : to_rename_array) {
-    if (_Externalize_Symbol(to_externalize, ExternalizationType::RENAME)) {
-      /* Update the function names for the ClosurePass.  */
-      to_externalize = RENAME_PREFIX + to_externalize;
-    }
+    SymbolsMap.insert({to_externalize, {"", ExternalizationType::RENAME, false, false}});
+    to_externalize = RENAME_PREFIX + to_externalize;
   }
+
+  TypeUpdaterVisitor(*this).TraverseDecl(AST->getASTContext().getTranslationUnitDecl());
```

Comment:
This line is quite cryptic for people that is not used to work with clang's AST. Please comment what it does.